[Navigation] Handle handler callback throwing an exception

rwlbuis opened the pull request:
#### 13f4ac88fbf3059b7323b8ee99274f350031691d
<pre>
[Navigation] Handle handler callback throwing an exception
<a href="https://bugs.webkit.org/show_bug.cgi?id=279061">https://bugs.webkit.org/show_bug.cgi?id=279061</a>

Reviewed by Chris Dumez.

Add IDL logic to treat exceptions being thrown from a callback as a returned promise rejected with that exception, in case the return type is a Promise [1].

The &quot;wait for all&quot; logic in Navigation.cpp knows how to deal with such rejected promises, so intercept-handler-throws.html passes.

[1] <a href="https://webidl.spec.whatwg.org/#js-invoking-callback-functions">https://webidl.spec.whatwg.org/#js-invoking-callback-functions</a> (Step 14.7)

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-crash-set-exception-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-from-invalid-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateCallbackImplementationContent):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
(WebCore::JSTestCallbackInterface::callbackThatRethrowsExceptions):
(WebCore::JSTestCallbackInterface::callbackThatTreatsExceptionAsRejectedPromise):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.h:
* Source/WebCore/bindings/scripts/test/TestCallbackInterface.idl:

Canonical link: <a href="https://commits.webkit.org/283560@main">https://commits.webkit.org/283560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e586ff290ae46cd8c45fdfa5d8cee6fa026c3e7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66611 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45986 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70645 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17744 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53388 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11970 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69678 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42339 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34042 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15034 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16098 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15375 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72347 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10568 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60705 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57714 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61034 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8696 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2317 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41793 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42870 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44053 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42613 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -6965,11 +6965,16 @@ sub GenerateCallbackImplementationContent
                 push(@$contentRef, "        auto throwScope = DECLARE_THROW_SCOPE(vm);\n");
                 push(@$contentRef, "        throwException(&lexicalGlobalObject, throwScope, returnedException);\n");
+                push(@$contentRef, "        return CallbackResultType::ExceptionThrown;\n");
+            } elsif ($operation->extendedAttributes->{TreatExceptionAsRejectedPromise}) {
```
I think we should match the spec instead:
https://webidl.spec.whatwg.org/#js-invoking-callback-functions

We shouldn't need a new `TreatExceptionAsRejectedPromise` non-standard IDL extended attribute. The spec says that if the callback's return type is a promise and an exception is thrown, we should reject the promise with this exception. So we should conditionalize your new behavior on whether or not the return type is a promise, NOT on a new `TreatExceptionAsRejectedPromise` non-standard IDL extended attribute I think. 

Hopefully, it's not that much more work and it would improve the spec compliance of our generated bindings.

rwlbuis replied:
Ah nice, I did not realise it was specced like that. I agree, nice to be compatible with the spec here.

cdumez commented on Source/WebCore/page/NavigationInterceptHandler.idl:
```diff
@@ -26,5 +26,6 @@
     GenerateIsReachable=ImplScriptExecutionContext,
-    SkipCallbackInvokeCheck
+    SkipCallbackInvokeCheck,
+    RethrowException
```
Could you clarify why we need RethrowException here? Seems odd that we need any custom IDL attribute to get standard behavior.

rwlbuis replied:
I misread the spec, I thought exceptionBehavior could be two values with a default of "rethrow",. but failed to read the sentence about the special handling when callable’s return type is a promise type, basically ignoring exceptionBehavior.

cdumez reviewed (approved):
LGTM if the bots are happy.

rwlbuis committed:
[Navigation] Handle handler callback throwing an exception
https://bugs.webkit.org/show_bug.cgi?id=279061

Reviewed by Chris Dumez.

Add IDL logic to treat exceptions being thrown from a callback as a returned promise rejected with that exception, in case the return type is a Promise [1].

The "wait for all" logic in Navigation.cpp knows how to deal with such rejected promises, so intercept-handler-throws.html passes.

[1] https://webidl.spec.whatwg.org/#js-invoking-callback-functions (Step 14.7)

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/no-crash-set-exception-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-get-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/transition-skipped-from-invalid-callback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateCallbackImplementationContent):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
(WebCore::JSTestCallbackInterface::callbackThatRethrowsExceptions):
(WebCore::JSTestCallbackInterface::callbackThatTreatsExceptionAsRejectedPromise):
* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.h:
* Source/WebCore/bindings/scripts/test/TestCallbackInterface.idl:

Canonical link: https://commits.webkit.org/283560@main
